src/main/java/org/threadly/concurrent/future/watchdog/MixedTimeWatchdog.java

```diff
@@ -0,0 +1,191 @@
+    this(scheduler, sendInterruptOnFutureCancel, DEFAULT_RESOLUTION_MILLIS);
+  }
+
+  /**
+   * Constructs a new {@link MixedTimeWatchdog} with a scheduler of your choosing.  It is critical 
+   * that this scheduler has a free thread available to inspect futures which may not have 
+   * completed in the given timeout.  You may want to use a org.threadly.concurrent.limiter to 
+   * ensure that there are threads available.
+   * <p>
+   * This constructor allows you to set the timeout resolutions.  Setting the resolution too large
+   * can result in futures timing out later than you expected.  Setting it too low results in 
+   * heavy memory consumption when used with a wide variety of timeouts.
+   * 
+   * @param scheduler Scheduler to schedule task to look for expired futures
+   * @param sendInterruptOnFutureCancel If {@code true}, and a thread is provided with the future, 
+   *                                      an interrupt will be sent on timeout
+   * @param resolutionMillis The resolution to allow timeout granularity
+   */
+  public MixedTimeWatchdog(SubmitterScheduler scheduler, boolean sendInterruptOnFutureCancel, 
+                           long resolutionMillis) {
+    ArgumentVerifier.assertGreaterThanZero(resolutionMillis, "resolutionMillis");
+    
+    this.scheduler = scheduler;
+    this.sendInterruptOnFutureCancel = sendInterruptOnFutureCancel;
+    cachedDogs = new ConcurrentHashMap<>();
+    watchdogProducer = (timeout) -> {
+      maybeScheduleCleaner();
+      return new ConstantTimeWatchdog(scheduler, timeout, sendInterruptOnFutureCancel);
+    };
+    cacheCleaner = new CleanRunner();
+    cleanerScheduled = new AtomicBoolean(false);
+    this.resolutionMillis = resolutionMillis;
+  }
+
+  /**
+   * Check how many futures are currently being monitored for completion.
+   * 
+   * @return The number of futures being monitored
+   */
+  public int getWatchingCount() {
+    int result = 0;
+    for (ConstantTimeWatchdog wd : cachedDogs.values()) {
+      result += wd.getWatchingCount();
+    }
+    return result;
+  }
+  
+  /**
+   * Watch a given {@link ListenableFuture} to ensure that it completes within the provided 
+   * time limit.  If the future is not marked as done by the time limit then it will be 
+   * completed by invoking {@link ListenableFuture#cancel(boolean)}.  Weather a {@code true} or 
+   * {@code false} will be provided to interrupt the running thread is dependent on how this 
+   * {@link MixedTimeWatchdog} was constructed.
+   * 
+   * @param timeoutInMillis Time in milliseconds that future should be completed within
+   * @param future Future to inspect to ensure completion
+   */
+  public void watch(long timeoutInMillis, ListenableFuture<?> future) {
+    long adjustedTimeout = timeoutInMillis / resolutionMillis; // int division to zero out to resolution
+    adjustedTimeout *= resolutionMillis;
+    if (adjustedTimeout != timeoutInMillis) {
```

Comment:
I feel like this might be a little wrong and should be 
```
if (adjustedTimeout < timeoutInMillis)
```
I also feel like this might optimize it a little more (maybe not would have to test it).
```
long adjustedTimeout = resolutionMillis
if (timeoutInMillis > adjustedTimeout) {
  adjustedTimeout = (long)(Math.ceil(timeoutInMillis /(double)resolutionMillis)*resolutionMillis);
}
```

Reply:
I have done some microbenchmarking using JMH to validate this as being the most optimal implementation.  I am pretty sure it's correct, we have unit testing around it too